### PR TITLE
Eliminate redundant Expr::addNulls calls

### PR DESCRIPTION
**Pull request by mbasmanova**
Summary:
Expr::addNulls can be expensive as it my require making vector writable by
copying it into a new vector. Hence, it is important to make sure not to call
it if there are no nulls to add. 

This change avoids calling Expr::addNulls unnecessarily by tightening checks for
new nulls in MutableRemainingRows::mayHaveChanged and Expr::removeSureNulls.
Rename MutableRemainingRows::mayHaveChanged to hasChanged to match new
behavior.

This optimization reduces expression evaluation CPU and wall time 2x for one of
the affected queries.

This change uncovered a bug in tranform_keys Presto lambda function. The check
for keys containing nulls was incorrectly skipped if there were no null keys. This change
includes a fix.

Differential Revision: D53264651

### CODE REVIEW

File: velox/functions/prestosql/TransformKeys.cpp [hunk 1]
```diff
@@ -80,7 +80,8 @@ class TransformKeysFunction : public exec::VectorFunction {
           elementToTopLevelRows,
           &transformedKeys);
 
-      if (transformedKeys->mayHaveNulls()) {
+      if (transformedKeys->mayHaveNulls() ||
+          transformedKeys->mayHaveNullsRecursive()) {
```
Comment (bikramSingh91) [inline velox/functions/prestosql/TransformKeys.cpp:84]: is this an unrelated bug fix?
Also, transformedKeys->mayHaveNullsRecursive() should cover transformedKeys->mayHaveNulls() as well so we only need mayHaveNullsRecursive()

Comment (mbasmanova) [reply velox/functions/prestosql/TransformKeys.cpp:84]: This fix was uncovered by the change. Without it a test fails.

I have copied this code from Map.cpp 

```
      // Verify there are no null keys.
      auto keysElements = keysArray->elements();
      if (keysElements->mayHaveNulls() ||
          keysElements->mayHaveNullsRecursive()) {
```

velox/functions/prestosql/MapFromEntries.cpp has similar:

```
    // Validate all map entries and map keys are not null.
    if (decodedRowVector->mayHaveNulls() || keyVector->mayHaveNulls() ||
        keyVector->mayHaveNullsRecursive()) {
```

Maybe we can look into simplifying all these in a follow up.

Comment (bikramSingh91) [reply velox/functions/prestosql/TransformKeys.cpp:84]: Thanks for the clarification, can you please add a short line in the commit message talking about it as its not obvious how its related to the change? Thanks!

**Review by bikramSingh91 — APPROVED** [review]

File: velox/functions/prestosql/TransformKeys.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/functions/prestosql/TransformKeys.cpp:84]: Added 

```
This change uncovered a bug in tranform_keys Presto lambda function. The check
for keys containing nulls was incorrectly skipped if there were no null keys. This change includes a fix.
```
